### PR TITLE
HEL-16 | Set up an automated build using CI

### PR DESCRIPTION
**Pull request by klempine**
I created an initial setup for Gitlab deployments. This PR contains:

- The endpoints `healthz` and `readiness` which Gitlab polls to see if the app is OK
- The Gitlab configuration file with some initial config for review, staging and production environments
- The script which will be run after each deployment. The script performs migration, admin user creation and initial page content creation

I refactored the original admin user creation which was in the Docker entrypoint script so that the `add_admin_user` command does the actual password generation instead of the script. This way there's slightly less code duplication.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        K8S_SECRET_VERSION: "$CI_COMMIT_SHORT_SHA"
+        K8S_SECRET_HKM_FEEDBACK_FROM_EMAIL: "no-reply@hel.ninja"
+
+production:
+    variables:
+        K8S_SECRET_ALLOWED_HOSTS: "*"
+        K8S_SECRET_DEBUG: 0
+        K8S_SECRET_SECRET_KEY: "$GL_STABLE_DJANGO_SECRET_KEY"
+        K8S_SECRET_VERSION: "$CI_COMMIT_SHORT_SHA"
+        K8S_SECRET_HKM_FEEDBACK_FROM_EMAIL: "no-reply@hel.fi"
```
Comment (charn) [inline .gitlab-ci.yml:37]: Uh oh, no new line at the end of the file. :scream:

Comment (klempine) [reply .gitlab-ci.yml]: I've hit enter now.

File: hkm/management/commands/add_admin_user.py [hunk 2]
```diff
@@ -10,14 +13,28 @@ def add_arguments(self, parser):
...
         )
         parser.add_argument(
             "-e", "--email", type=str, help="Email", default="admin@example.com"
         )
 
     def handle(self, *args, **kwargs):
         if not get_user_model().objects.filter(username=kwargs["username"]).count():
+            password = kwargs["password"]
+            if not password:
+                password = self.random_string(20)
```
Comment (charn) [inline hkm/management/commands/add_admin_user.py:26]: Just FYI, there seems to be [a built-in way](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#django.contrib.auth.models.BaseUserManager.make_random_password) to accomplish something similar:

```python
from django.contrib.auth import get_user_model
User = get_user_model()
password = User.objects.make_random_password(20)
```

File: .prod/on_deploy.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+echo "on_deploy.sh here ..."
+
+python /app/manage.py migrate --noinput
+
+# Generate the admin user using the password given in the environment variables.
+# If no password is set, the admin user gets a generated password which will
+# be written in stdout so that it can be accessed during the initial deployment.
+if [[ "$ADMIN_USER_PASSWORD" ]]; then
```
Comment (charn) [inline .prod/on_deploy.sh:10]: I this the admin creation here because we can't use `CREATE_ADMIN_USER` on the `docker-entrypoint.sh` in review/staging/production?

File: .prod/on_deploy.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    python /app/manage.py add_admin_user -u admin -p $ADMIN_USER_PASSWORD -e admin@example.com
+else
+    python /app/manage.py add_admin_user -u admin -e admin@example.com
+fi
+
+python /app/manage.py create_initial_page_content
```
Comment (charn) [inline .prod/on_deploy.sh:16]: I wonder if this will cause problems in staging/production like persistent environments since the script is doing: 

`PageContent.objects.create` instead of e.g. `PageContent.objects.get_or_create`?

Comment (charn) [reply .prod/on_deploy.sh:16]: Okay, my bad, I was checking things from `master` branch, which was not the right branch to check from. Everything seems to be in order, please carry on.

File: kuvaselaamo/wsgi.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
-# -*- coding: utf-8 -*-
```
Comment (charn) [inline kuvaselaamo/wsgi.py:1]: I guess this should be fine to remove here even though it's still a Python 2.7 project.

File: .prod/on_deploy.sh (hunk 3 above)
Comment (klempine) [reply .prod/on_deploy.sh:10]: I've been wondering about this myself. We can use the env variables and `docker-entrypoint.sh`, unless I'm mistaken.

The current setup is about the same as in Helsinki profile which means that `on_deploy.sh` and `docker-entrypoint.sh` are doing basically the same thing, except that `docker-entrypoint.sh` can be controlled more with env variables.

Initially I thought that Gitlab uses only `on_deploy.sh`, but I think it's also executing `docker-entrypoint.sh`.

Do you have a suggestion on how this should be? Should I just remove the duplicate stuff from `on_deploy.sh` and put suitable env vars (e.g. `CREATE_ADMIN_USER = 1`) in the Gitlab variables so that `docker-entrypoint.sh` is used to migrate the database and create admin/page contents?

File: hkm/management/commands/add_admin_user.py (hunk 2 above)
Comment (klempine) [reply hkm/management/commands/add_admin_user.py:26]: Thanks for the tip, I was unaware of this + my question on #kuva-aok-development didn't get any responses.

I'll remove my custom stuff and replace it with your suggestion.

File: .prod/on_deploy.sh (hunk 4 above)
Comment (klempine) [reply .prod/on_deploy.sh:16]: this-is-fine

File: hkm/management/commands/add_admin_user.py (hunk 2 above)
Comment (klempine) [reply hkm/management/commands/add_admin_user.py]: I've pushed the change now.

File: kuvaselaamo/wsgi.py (hunk 5 above)
Comment (klempine) [reply kuvaselaamo/wsgi.py:1]: That one was there previously until I replaced the file contents with Helsinki profile's code. I didn't even notice it until now.

If you think that the line is relevant, I can put it back.

Comment (charn) [reply kuvaselaamo/wsgi.py:1]: There's no UTF-8 strings on that file, so you don't have to put it back.

File: .prod/on_deploy.sh (hunk 3 above)
Comment (charn) [reply .prod/on_deploy.sh:10]: I think `on_deploy.sh` is the correct place to do migrations on review/staging/production. I think that file is guaranteed to be run only once. I guess theoretically docker-entrypoint might run at the same time on multiple containers (if the service was scaled to run e.g. 3 containers) which might have interesting results if a migration script for that service was started at the same time an all containers.

TL;DR I think `on_deploy.sh` is fine for now. :+1:

File: docker-entrypoint.sh [hunk 6]
```diff
@@ -41,3 +38,5 @@ elif [[ "$DEV_SERVER" = "1" ]]; then
 else
     uwsgi --ini .prod/uwsgi.ini
 fi
+
+echo "docker-entrypoint.sh here ... done"
```
Comment (charn) [inline docker-entrypoint.sh:42]: I think this line will be printed when the container stops, I think. But maybe that's what we want.

File: .prod/on_deploy.sh (hunk 3 above)
Comment (klempine) [reply .prod/on_deploy.sh:10]: OK, we'll go with that, thanks!

File: docker-entrypoint.sh (hunk 6 above)
Comment (klempine) [reply docker-entrypoint.sh:42]: Just some debug printing, I'll probably remove it at some point.
